hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,244 @@
+import org.apache.hadoop.ozone.protocol.commands.ReplicateContainerCommand;
+import org.apache.hadoop.ozone.protocol.commands.SCMCommand;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Handles the EC Under replication processing and forming the respective SCM
+ * commands.
+ */
+public class ECUnderReplicationHandler implements UnderReplicationHandler {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(ECUnderReplicationHandler.class);
+  private final ECContainerHealthCheck ecContainerHealthCheck =
+      new ECContainerHealthCheck();
+  private final PlacementPolicy containerPlacement;
+  private final long currentContainerSize;
+  private final NodeManager nodeManager;
+
+  public ECUnderReplicationHandler(
+      final PlacementPolicy containerPlacement, final ConfigurationSource conf,
+      NodeManager nodeManager) {
+    this.containerPlacement = containerPlacement;
+    this.currentContainerSize = (long) conf
+        .getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+            ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.nodeManager = nodeManager;
+  }
+
+  /**
+   * Identify a new set of datanode(s) to reconstruct the container and form the
+   * SCM command to send it to DN. In the case of decommission, it will just
+   * generate the replicate commands instead of reconstruction commands.
+   *
+   * @param replicas - An instance of ContainerReplicaCount, containing the
+   *                   current replica count and inflight adds and deletes
+   * @param pendingOps - Inflight replications and deletion metrics.
+   * @param result - Health check result.
+   * @param remainingMaintenanceRedundancy - represents that how many nodes go
+   *                                      into maintenance.
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final ContainerReplicaPendingOps pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+    List<ContainerReplicaOp> pendingOpsForContainer =
+        pendingOps.getPendingOps(container.containerID());
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas, pendingOpsForContainer,
+            remainingMaintenanceRedundancy);
+
+    ContainerHealthResult.UnderReplicatedHealthResult containerHealthResult =
```

Comment:
Is it safe to cast this directly, without first checking if the result is underReplicated? I am not sure what will happen if, for example, the result is OverReplicated - I suspect it will give some sort of cast / runtime error.

I think we need to check the result:

```
ContainerHealthResult currentHealth = ecContainerHealthCheck
            .checkHealth(container, replicas, pendingOpsForContainer,
                remainingMaintenanceRedundancy);
if (currentHealth.getHealthState() != UNDER_REPLICATED) {
  // Log and return null
}
ContainerHealthResult.UnderReplicatedHealthResult currentUnderRepRes = ((ContainerHealthResult.UnderReplicatedHealthResult) currentHealth);
```

Then we can simplify the condition at line 109 a little, as we know for sure the state is UNDER_REP so we don't need to check it there. Similar for the condition at 174 - we can just use the already cast variable from here.